### ISSUE TITLE
DiCE factors from importance sampling?

### ISSUE THREAD

**Issue by mattwescott** [pyro-ppl/funsor#151]
In `Tensor.unscaled_sample`, a DiCE factor is added to the sample. In `Gaussian.unscaled_sample`, this isn't necessary because of the parameterization of the delta function. 

So far including DiCE with the samples doesn't sit quite right conceptually. Could it make sense to instead approach this from importance sampling? The default interpretation of `adjoint(p) @ f` for continuous p might be to use importance sampling with some default proposal q. To maintain differentiability, the importance weights would be `p / q.detach()`. If samples can be drawn from p, then p can be used for the proposal, and the importance weights would be `p / p.detach()`, which in this context are the DiCE factors.

Am I missing some benefits of the DiCE formalism?

**Comment by fritzo** [comment]
Hi @mattwescott, my aim of including the dice factor in non-reparametrizable `Tensor.sample()` is to maintain the invariant that samples are unbiased at all derivatives. That is, for any funsor $p(theta)$ and C-infinity test function `f(theta)` both depending on a parameter `theta`,
```
E[ p(theta).sample() * f(theta) ] == E[ p(theta) * f(theta) ]
                                at all
                              derivatives
                               wrt theta
```
Thus we can think of `.sample()` as a differentiable approximation, in particular as an approximation by `Delta` funsors. I find it conceptually cute to define 'sampling' as exactly 'approximation by Delta funsors'.

Re: importance sampling, I guess the first way I see that fitting into Funsor is via yet another interpretation: whereas the `monte_carlo` interpretation samples from the prior, `importance_monte_carlo` (or maybe `monte_carlo(proposal=q)`) could sample from another distribution. Does that sound reasonable to you?

**Comment by mattwescott** [comment]
Thanks @fritzo. Unless I'm still confused, what I have in mind might generalize this. 

It seems natural to have a `differentiable_importance_sample(proposal=q)` method that would have the properties you've outlined of preserving expectations of all derivatives. The current behavior of `funsor.sample` would be a special case of using p as its own proposal. 

```python
class Funsor:

    def differentiable_importance_sample(self, proposal=None):
        if proposal is None:
            proposal = self
        sample = proposal.raw_sample()
        return reduce(add, [Delta(sample), proposal.sum(), self(sample)/proposal(sample).detach()])

    def sample(self):
        # However it is implemented, this definition will take some adjustment for me
        return self.differentiable_importance_sample()
```

This feels like a simpler and more general way to conceptualize DiCE, but I might still be missing something. 

> I find it conceptually cute to define 'sampling' as exactly 'approximation by Delta funsors'.

This will probably grow on me.

**Comment by eb8680** [comment]
@mattwescott we prefer when possible to build more complex machinery out of simpler pieces in Funsor.  Dice is a nice atomic piece of functionality; to estimate an integral with importance sampling we'd just write out the definition: 
```python
# monte carlo
with interpretation(monte_carlo):
    y = funsor.Integrate(logp, f)

# importance sampling
with interpretation(monte_carlo):
    y = funsor.Integrate(logq, f * (logp - logq).exp())
```

**Comment by mattwescott** [comment]
Thanks @eb8680, I'm getting closer to the same page. Conceptually I found it easier to motivate Dice from importance sampling. But as you said, however it is arrived at, it is a nice atom of functionality and the implementation probably shouldn't be entangled with importance sampling. 

If I could bother you to comment on another aspect of this: Would it be practical to move the Dice logic to a base class? Maybe the parameterization could be handled with something like this?

```python

def detach(fn, params):
    """this poutine turns off gradients for some parameters"""
    return detached_fn

class Funsor:
    def sample(self):
        sample = self.raw_sample()
        if not self.fully_reparameterized:
            dice_factor = detach(self, self.reparameterized_params)(sample)/self(sample).detach()
        else:
            dice_factor = 0.0
        return reduce(add, [Delta(sample), dice_factor, self.sum()])
```

**Comment by fritzo** [comment]
Again, I prefer to think of `raw_sample()` as low-level concept appropriate for FORTRAN/MATLAB/CUDA code, and to think of `.sample()` as the more mathematical concept, an atomic method that each `Funsor` subclass must implement. Note also there are partially reparametrized cases where it doesn't make sense to consult a boolean `.fully_reparametrized` flag; even TFP made a decision early on to allow more than two values to `.reparametrization_type`.

**Comment by mattwescott** [comment]
Thanks for the detail @fritzo, that is even clearer now. It actually wasn't my intention to readjudicate `.sample()`. I'm curious if it would be practical to factor out the Dice factor computation.

> Note also there are partially reparametrized cases where it doesn't make sense to consult a boolean .fully_reparametrized

Using `.reparametrization_type` seems cleaner, switched the pseudocode. 

For the partial case, my hypothetical (and quite possibly nonsensical) `partial_detach` poutine turns off gradients for the relevant subset of parameters. These parameters are hypothetically stored in `.reparam_params` 

```python
def partial_detach(fn, params):
    """this poutine turns off gradients for params"""
    ...
    return partially_detached_fn

def dice_factor(p, raw_sample):
    if p.reparametrization_type == 'full':
        return 0.0
    elif p.reparametrization_type == 'none':
        return p(raw_sample) - p(raw_sample).detach()
    elif p.reparametrization_type == 'partial':
        return partial_detach(p, p.reparam_params)(raw_sample) - p(raw_sample).detach()
    else:
        raise "invalid reparametrization type"
```

Seems like it could be expensive to reevaluate the density with different `.requires_grad`. It is not clear to me how that would work, both the funsor evaluation and torch internals are still opaque. If it is expensive, could be a case for keeping the Dice factor computation with the sampling code where the parameterization is explicit?

Thanks for taking the time, it has been helpful.

**Comment by eb8680** [comment]
@mattwescott I don't quite understand the motivation for your proposed refactoring - do you want to be able to turn off Dice?  When would you want a `funsor.Delta` that would silently produce incorrect gradient estimators?

**Comment by mattwescott** [comment]
> I don't quite understand the motivation for your proposed refactoring

My motivation is mostly to better understand Funsor. By proposing what seems to me a simplification, my hope is that if it is misguided you might point me toward what I'm missing. And if it could make sense and simply isn't a priority, that is helpful to know as well.

The implementations of `.unscaled_sample()` seemed complex. It would have felt more natural to me if any reparametrization were an explicit property of the distribution, with the Dice factor being constructed from the parametrization. 

The pseudocode in my last comment might not be practical, especially for some cases of partial reparametrization. If instead the stochastic source and transformation are both explicit, both `unscaled_sample` and `compute_dice_factor` might be even simpler:

```python
def compute_dice_factor(stochastic_source, epsilon):
    return stochastic_source(epsilon) - stochastic_source(epsilon).detach()

class Distribution(Funsor):
    def unscaled_sample(self):
        epsilon = self.stochastic_source.raw_sample()
        dice_factor = compute_dice_factor(stochastic_source, epsilon)
        point = self.transformation(epsilon)
        return reduce(add, [Delta(point), dice_factor, self.sum()])
```

In any case, a factored out Dice might be used by all the `.sample()` methods, and could be easier to understand and test. My primary interest though is not to advocate for these changes, but to hopefully find out if there is something I'm not considering.

**Comment by eb8680** [comment]
Hmm, interesting.  Given that there's currently only one implementation of Dice in Funsor (in `Tensor.unscaled_sample`) I don't quite see how your proposal would simplify the existing code, though it might more directly reflect one mathematical presentation of Dice, as you say.  

Note that we have no plans to reimplement any additional samplers, as there are already high-performance, fully reparameterzed, broadcasting-aware samplers available for most common continuous-valued distributions in `torch.distributions` or `pyro.distributions` that we would prefer to wrap when possible.  

If you think a version of your proposal would make it easier to do that (e.g. by reusing [Pyro's `score_parts` machinery](http://docs.pyro.ai/en/stable/distributions.html#pyro.distributions.Distribution.score_parts) for Dice), feel free to submit a PR adding `unscaled_sample` methods to some of the classes in `funsor/distributions.py` and we can discuss it in detail there.

**Comment by mattwescott** [comment]
> e.g. by reusing Pyro's score_parts machinery for Dice where appropriate

Leaving `score_parts` behind is part of the appeal for me of explicit reparametrization. I've been seeing `score_parts` as loosely tied to parameterization since a comment about possibly using it to [stick the landing](https://arxiv.org/abs/1703.09194). Maybe I got the wrong impression. In any case, leaning on `score_parts.score_function` for now to compute Dice factors would make engineering sense to me.

Will go ahead and close this. Thanks for all the discussion.
